pyteal/ast/subroutine.py

```diff
@@ -49,15 +55,32 @@ def __init__(
                     )
                 )
 
+            expected_arg_type: type = Expr
+            if name in implementation.__annotations__:
```

Comment:
@tzaffi To test my understanding:  Is there a situation where `name` won't be in `__annotations__`?

If `name` is always present, I'm then wondering what requires using `signature` and `__annotations__`?

Reply:
Name won't show up in __annotations__ in the case that defined subroutine didn't have an explicit annotation, and in that case we default to Expr

Reply:
@tzaffi To fill in the gap in my intuition:  What prevents (or makes undesirable) deferring the validation until `evaluateSubroutine`?

* As I understand it, the `SubrouteDefinition` constructor is forced to assume a parameter type in order to build `expected_arg_types` and `by_ref_args`.  Obviously, without a type parameter hint, a _guess_ is made:  `Expr`.
* Then, during `invoke` the _guess_ is checked against the provided argument and aborts evaluation when the _guess_ is inaccurate.
* I started down the path because I didn't immediately grok _why_ `Expr` is chosen as the default.  With the above understanding in mind, I'm wondering if there's more intuition I missed that explains why validation happens in the constructor with partial information.
  * My _assumption_ is:  Enforcing `Expr` as the only supported type without annotation simplifies life for PyTeal maintainers.  If that's the case, it works for me though I want to verify.

Reply:
> @tzaffi To fill in the gap in my intuition: What prevents (or makes undesirable) deferring the validation until `evaluateSubroutine`?
In general, it's good to "fail fast" so there would have to be a strong argument (e.g. an object is invalid because it can be made valid later) to defer validation to a later stage.
> 
> * As I understand it, the `SubrouteDefinition` constructor is forced to assume a parameter type in order to build `expected_arg_types` and `by_ref_args`.  Obviously, without a type parameter hint, a _guess_ is made:  `Expr`.
It's more than a "guess". We have just introduced subroutine annotations in #182, so by default parameter types should be backwards compatible, and therefore an `Expr`.


> * Then, during `invoke` the _guess_ is checked against the provided argument and aborts evaluation when the _guess_ is inaccurate.
If we view this not as a "guess" but as an "implicit declaration" does aborting evaluation sound less onerous?

> * I started down the path because I didn't immediately grok _why_ `Expr` is chosen as the default.  With the above understanding in mind, I'm wondering if there's more intuition I missed that explains why validation happens in the constructor with partial information.
The reason is "backwards compatibility".

>   
>   * My _assumption_ is:  Enforcing `Expr` as the only supported type without annotation simplifies life for PyTeal maintainers.  If that's the case, it works for me though I want to verify.
alas, that simplified life as about to get more complicated with the upcoming introduction of ABI type params

Reply:
@tzaffi Thanks for the explanation - makes sense to me.  

Since I spent some time understanding the topic, I thought an inline comment _may_ be beneficial.  I opened an optional child PR:  https://github.com/algorand/pyteal/pull/217.

If you don't feel the PR is an improvement worth taking in, I'll happily close it.  Either way, I learned a bit more about PyTeal.